Dashboard: Alerts user to incorrect tag format for JSON import

iamelDuderino opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

Fixes #54285: Malformed tags cause hidden title and settings page crash

**Which issue(s) this PR fixes**:
Fixes #54285
<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

**Special notes for your reviewer**:

<img width="644" alt="image" src="https://user-images.githubusercontent.com/13452266/188192973-340728a6-0f34-4037-9a64-07ea727f3d56.png">

<img width="647" alt="image" src="https://user-images.githubusercontent.com/13452266/188193057-8a68dbf4-c09b-4510-8b0c-b00cf8741cfd.png">

iamelDuderino committed:
Dashboard: Alerts user to incorrect tag format for JSON import

Fixes #54285: Malformed tags cause hidden title and settings page crash

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=54657) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=54657) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=54657) it.</sub>

polibb commented on public/app/features/manage-dashboards/utils/validation.ts:
```diff
@@ -4,8 +4,29 @@ import { validationSrv } from '../services/ValidationSrv';
+    }
   } catch (error) {
     return 'Not valid JSON';
   }
```
```suggestion
  }
  if (dashboard.tags) {
    if (Array.isArray(dashboard.tags)) {
      const hasInvalidTag = dashboard.tags.some((tag: string) => typeof tag !== 'string');
      if (hasInvalidTag) {
        return 'error: tags expected Array of Strings';
      }
    } else {
      return 'error: tags expected Array';
    }
  }
      
```

polibb replied:
Having less `if else`s and variables makes it easier to read IMO. What do you think of those suggestions?

polibb commented on public/app/features/manage-dashboards/utils/validation.ts:
```diff
@@ -4,8 +4,29 @@ import { validationSrv } from '../services/ValidationSrv';
   try {
-    JSON.parse(json);
-    return true;
+    let dashboard = JSON.parse(json);
+    // If dashboard.tags provided && is Array
+    if (dashboard.tags && Array.isArray(dashboard.tags)) {
+      // Check dashboard.tags in Array for only Strings
+      let cont = true;
+      dashboard.tags.forEach((tag: string) => {
+        if (cont === true && typeof tag !== 'string') {
+          cont = false;
+        }
+      });
+      // Continue to import dashboard if Tags in correct format
+      if (cont) {
+        return true;
+      } else {
+        return 'error: tags expected Array of Strings';
+      }
+      // If dashboard.tags && not Array
+    } else if (dashboard.tags && !Array.isArray(dashboard.tags)) {
+      return 'error: tags expected Array';
+      // Else generic JSON test passes and no Tags provided, then import dashboard
+    } else {
+      return true;
+    }
```
```suggestion
    let dashboard = JSON.parse(json);
```

polibb replied:
It seems that there's no need of this code block to be in a `try {} catch {}`. We can just leave the `JSON.parse` then and not return true until all validation has passed?

ivanortegaalba replied:
If I'm not wrong, the try/catch is part of the validation. An invalid JSON can be introduced in the input, and this function will be called with it. If we don't have the try/catch the function will throw an uncontrolled error.

polibb replied:
When an invalid JSON is introduced in the input, the try/catch will catch the error and return `return 'Not valid JSON';`. But the rest of the validation in this method has no need of a try/catch since it's not going to throw unexpected errors from anywhere.

polibb commented on public/app/features/manage-dashboards/utils/validation.ts:
```diff
@@ -4,8 +4,29 @@ import { validationSrv } from '../services/ValidationSrv';
-    JSON.parse(json);
-    return true;
+    let dashboard = JSON.parse(json);
+    // If dashboard.tags provided && is Array
```
The comments don't add any new information and the code is easy to read, so what do you think about having no comments? It seems to just create twice as many lines

polibb reviewed (changes requested):
Thank you for including this validation and tests for it! 
I haven't seen snake case anywhere in the tests, why did you choose it over camel case here?

iamelDuderino committed:
Update public/app/features/manage-dashboards/utils/validation.ts

Co-authored-by: Polina Boneva <13227501+polibb@users.noreply.github.com>

polibb commented on public/app/features/manage-dashboards/utils/validation.ts:
```diff
@@ -4,11 +4,44 @@ import { validationSrv } from '../services/ValidationSrv';
 
 export const validateDashboardJson = (json: string) => {
   try {
```
```suggestion
  let dashboard;
  
  try {
```

polibb commented on public/app/features/manage-dashboards/utils/validation.ts:
```diff
@@ -4,11 +4,44 @@ import { validationSrv } from '../services/ValidationSrv';
-    JSON.parse(json);
-    return true;
+    let dashboard = JSON.parse(json);
+    // If dashboard.tags provided && is Array
+    if (dashboard.tags && Array.isArray(dashboard.tags)) {
+      // Check dashboard.tags in Array for only Strings
+      let cont = true;
+      dashboard.tags.forEach((tag: string) => {
+        if (cont === true && typeof tag !== 'string') {
+          cont = false;
+        }
+      });
+      // Continue to import dashboard if Tags in correct format
+      if (cont) {
+        return true;
+      } else {
+        return 'error: tags expected Array of Strings';
+      }
+      // If dashboard.tags && not Array
+    } else if (dashboard.tags && !Array.isArray(dashboard.tags)) {
+      return 'error: tags expected Array';
+      // Else generic JSON test passes and no Tags provided, then import dashboard
```
```suggestion
```

polibb commented on public/app/features/manage-dashboards/utils/validation.ts:
```diff
@@ -4,11 +4,44 @@ import { validationSrv } from '../services/ValidationSrv';
   } catch (error) {
     return 'Not valid JSON';
   }
+  if (dashboard.tags) {
```
```suggestion
  if (dashboard && dashboard.tags) {
```

polibb commented:
Note for the future:
There is a great new way of generating models for the frontend and backend, a _schematisation_ of sorts. This is still behind feature flags (e.g. `validateDashboardsOnSave`) at the moment and is not known when we will have it done _not_ in alpha/beta state 😊. When it is done the model of the dashboard itself will correctly check for the contents of the JSON and this validation will probably be able to be removed. Right now it is much needed, nevertheless.

iamelDuderino committed:
Included Suggestions

- Removed Comments
- Updated Code Block accordingly
- Updated Tests to camelCase over snake_case

polibb commented on public/app/features/manage-dashboards/utils/validation.ts:
```diff
@@ -3,12 +3,18 @@ import { getBackendSrv } from '@grafana/runtime';
-    return true;
-  } catch (error) {
-    return 'Not valid JSON';
+  let dashboard = JSON.parse(json);
```
hi @iamelDuderino ✨. We need to pass the `JSON.parse(json)` through a `try {} catch {}` because if it fails the `validateDashboardJson` will throw an unexpected that's never handled and the page will fail to load.

iamelDuderino replied:
Hey @polibb Oops! You are correct in that the page just simply does nothing if the json is invalid. I have re-added the `try {} catch {}` and included a test for it, as well as reworded my initial errors to better align with the existing format from the catch statement. Please let me know if there is anything further, thanks!

iamelDuderino committed:
Updates per comments

- Re-wrapped function in try{}, catch{} as I appear to have overlooked including it in the initial refactor
- Re-worded errors to align with initial error
- Added a test case for invalid json

iamelDuderino committed:
Update validation.ts

Updated errors to read correctly to the root cause.
Updated dashboard variable as const.

iamelDuderino committed:
Update actions.test.ts

Fix tests according to error output rewording

iamelDuderino committed:
Update validation.ts

- Included test for an empty string of non-array

iamelDuderino committed:
Update actions.test.ts

-- Commented incorrect commit for validation.ts, update:
- Refactored code to better align and separate from generic JSON package tests followed by our manual checks of (1) Is array, and (2) if array, is of strings

- Test cases now include a check for non-array empty string in the tag property

polibb reviewed (commented):
Hello @iamelDuderino 😊. The solution looks good! 
It seems like the build is failing, would you try merging your branch with `grafana:main` to be up to date and to re-run the build, so we see if it's some flaky tests or something we have to look into and fix here? This is also probably going to fix the failing Enterprise build.

iamelDuderino committed:
Merge branch 'grafana:main' into 54285-fixes-malformed-tags-crash

iamelDuderino committed:
Merge branch 'grafana:main' into 54285-fixes-malformed-tags-crash

polibb reviewed (approved):
Great choice for a first task and impressive execution and implementation, @iamelDuderino!
